[MRG] add taxonomy subcommand

bluegenes opened the pull request:
This PR adds the top-level framework for a `sourmash tax` subcommand, and focuses on reading in gather results files for summarization or classification.

- `sourmash tax metagenome` - reads in one or more metagenome gather results files and taxonomies; summarizes at all ranks. Reports this as `csv_summary`, `krona` format at a chosen rank, or `lineage summary` at chosen rank, or any combination of these outputs. 
- `sourmash tax genome` - reads in one or more genome gather results files and taxonomies; classifies to "best" match given desired threshold or rank. Reports this as `csv`, `krona` format, or both.
- ` sourmash tax annotate` reads in one or more gather results files and taxonomies and adds a `lineage` column to each gather csv.

for all, writing to stdout is supported if only one output type is specified.

@bluegenes `to do` checklist: 
- [x] with https://github.com/dib-lab/sourmash/pull/1565, we should be able to read in more than one gather file at once for either `summarize` or `classify`, without needing to separately collect information about the query name. 
  - [x] This also means that we can output the combined lineage/sample summary file directly from `summarize` if desired. I will work on implementing this, unless there are thoughts otherwise.
- [x] find appropriate location for argparse threshold input `range_limited_float_type`
    - [ ]  this is implicitly tested - should it be tested directly? Where are `cli/utils.py` functions typically tested?
- [x] add function docstrings
- [x] integrate #1574
- [x] integrate hannah's summarize testing #1560 
- [x] properly restrict function args with *
- [x] basic documentation for new commands
- [x] tutorial / example usage -- see https://github.com/bluegenes/2021-sourmash-taxonomy-hackathon/blob/main/tax_demo.ipynb
- [x] gather results  - check for essential columns (and test bad gather inputs)
- [x] taxonomy - return available_ranks and check
  - if we can return `available_ranks` from `load_taxonomy_assignments`, then we can check that the desired classification rank is available within the lineage data (currently we just assume it's available). This would also enable classification at strain if/when available, with all the caveats that might bring about strains available in the database.
e.g. add this check to `classify` after loading `tax_assign`:
    ```
    if args.rank not in available_ranks:
          notify(f"No taxonomic information at rank {args.rank}: cannot classify at this rank")
    ```
choices, choices:
- [x] remove `combine`
- [x] rename `summarize` --> `metagenome`; `classify` --> `genome`; `label` --> `annotate` (but alias old names :)

**suggestions for 6/18 taxonomy hackathon:**
- improve tax documentation? or write some recipes?
- CAMI Output!? see https://github.com/luizirber/2020-cami/blob/master/scripts/gather_to_opal.py

design choices:

- for `summarize` and `classify`, we offer several optional outputs. I enabled this by using an `--output-base` parameter and internally adding standard extensions, e.g. `.summarized.csv` or `.krona.csv`. I am happy to change this if there's a better design / one more aligned with our other commands. The goal was to enable multiple output formats at once, and allow us to (relatively easily) enable additional output formats for visualization.
- `classify` needs to make some decisions to report the _best_ annotation. I've currently enabled doing this at either a desired rank or given a desired `threshold`. The benefit of the threshold is that we can go up a rank if there's not sufficient support for an annotation at that rank. 

@ctb @taylorreiter @luizirber thoughts?

ctb committed:
provide a kind of ridiculous upgrade to lca index to better deal with identifiers/taxonomy

ctb committed:
update load_taxonomy_assignments to be more flexible and pay attention to CLI

bluegenes committed:
init structure for taxonomy subcommand

bluegenes committed:
more init

bluegenes committed:
syntax and add tax to init

bluegenes committed:
Merge branch 'latest' into add-taxonomy

ctb committed:
Merge branch 'latest' into update/lca_index

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes committed:
init tax tests

HTML-CODDER reviewed (changes requested):
[WIP] framework for taxonomy subcommand #1543

bluegenes committed:
Merge branch 'latest' into update/lca_index

bluegenes committed:
Merge branch 'update/lca_index' of https://github.com/dib-lab/sourmash into add-taxonomy

bluegenes committed:
working tax summarize command

bluegenes committed:
fix main

bluegenes committed:
init tests for new tax_utils

bluegenes committed:
add ascending taxlist

bluegenes committed:
init classify cmd

bluegenes committed:
init tax cli testing

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes committed:
fix filename

bluegenes committed:
Merge branch 'add-taxonomy' of https://github.com/dib-lab/sourmash into add-taxonomy

luizirber commented on src/sourmash/cli/tax/classify.py:
```diff
@@ -0,0 +1,64 @@
+import sourmash
+from sourmash.logging import notify, print_results, error
+
+#https://stackoverflow.com/questions/12116685/how-can-i-require-my-python-scripts-argument-to-be-a-float-between-0-0-1-0-usin
```
I think this might be a simpler approach (and also easier to read/understand in the future? =])
https://stackoverflow.com/questions/55324449/how-to-specify-a-minimum-or-maximum-float-value-with-argparse#55410582

bluegenes committed:
change to function for classify threshold

bluegenes committed:
add header

bluegenes committed:
enable single gather result for summarize; mult for classify

taylorreiter committed:
add util script to take output of tax and format for krona viz (#1559)

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes committed:
get summarized working for summary and krona output

bluegenes committed:
init test krona output

bluegenes committed:
add write_summary function

bluegenes committed:
get classify working again, both summary and krona output

bluegenes committed:
test write_classification

bluegenes committed:
init classify cli tests

bluegenes committed:
init tests for load_taxonomy_assignments

bluegenes committed:
enable force for getting past duplicated entries in taxonomy csv

bluegenes committed:
handle and test missing taxonomy info

bluegenes committed:
classify: handle, test empty gather results, gather results from csv

bluegenes committed:
split identifiers by default

bluegenes committed:
standardize spacing

bluegenes committed:
comments

bluegenes committed:
init add tax to docs

bluegenes committed:
Merge branch 'add-taxonomy' of https://github.com/dib-lab/sourmash into add-taxonomy

taylorreiter committed:
[MRG] add a function to take multiple sourmash tax summarize csvs and output a single "abundance" df (#1562)

* add a function to take multiple sourmash tax summarize csvs and output a single 'abundance' df

* add import dep.

bluegenes committed:
rework format_tax_to_frac for easier testing and use; add tests

bluegenes committed:
better name and docstring for agg_sumgather_csvs_by_lineage

bluegenes committed:
init combine command

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes committed:
Merge branch 'latest' into add-taxonomy

ctb committed:
debugging code to help track down SBT duplicates/loss problem

ctb committed:
fix, I think

ctb committed:
remove unnecessary code

ctb committed:
add test for duplicate signatures in SBT creation

ctb committed:
Merge branch 'debug/sbt_dups_tests' into debug/sbt_dups

ctb committed:
see what happens when you run twice

ctb committed:
add missing signatures, oops

ctb committed:
initial refactoring that passes many tests

ctb committed:
factor filename generation out of actual writing

ctb committed:
refactor and cleanup

ctb committed:
add more sigs to test, add note of concern L)

ctb committed:
fix --append tests, too

ctb committed:
refactor out save_exact in favor if save(..., overwrite=True)

ctb committed:
fix some storage stuff in the tests

ctb committed:
make test less confusing?

ctb committed:
Update src/sourmash/sbt_storage.py

Co-authored-by: Luiz Irber <luizirber@users.noreply.github.com>

ctb committed:
define list_sbts() on base Storage class

ctb committed:
Merge branch 'debug/sbt_dups' of github.com:dib-lab/sourmash into debug/sbt_dups

ctb committed:
Merge branch 'latest' of https://github.com/dib-lab/sourmash into update/lca_index

ctb committed:
Merge branch 'debug/sbt_dups' into update/lca_index

ctb committed:
properly record duplicate signature names

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes commented:
This PR adds the top-level framework for a `sourmash tax` subcommand, and focuses on reading in gather results files for summarization or classification.

- `sourmash tax summarize` - reads in one gather results file; summarizes at all ranks. Reports this as `csv`, `krona` format at a chosen rank, or `lineage summary` at chosen rank, or any combination of these outputs. 
- ` sourmash tax combine` reads in one or more outputs of `summarize` and outputs a `csv` or `tsv` summary by lineage and sample.
- `sourmash tax classify` - reads in one or more gather results and associated query names; classifies to "best" match given desired threshold or rank. Reports this as `csv`, `krona` format, or both.

for all, writing to stdout is supported if only one output type is specified.

still to do: 
- [x] with https://github.com/dib-lab/sourmash/pull/1565, we should be able to read in more than one gather file at once for either `summarize` or `classify`, without needing to separately collect information about the query name. 
  - [x] This also means that we can output the combined lineage/sample summary file directly from `summarize` if desired. I will work on implementing this, unless there are thoughts otherwise.
- [ ] finish fleshing out, testing, checking, etc for `combine`.
- [x] find appropriate location for argparse threshold input `range_limited_float_type`
    - [ ]  this is implicitly tested - should it be tested directly? If yes, add test file for `cli/utils.py`...
- [ ] add documentation for new commands
- [ ] add function docstrings
- [ ] integrate #1574
- [ ] integrate hannah's summarize testing #1560 
- [ ] CAMI output format? see https://github.com/luizirber/2020-cami/blob/master/scripts/gather_to_opal.py
- [ ] better/safer `load_gather_results` function?
- [x] properly restrict function args with *

design questions:

- for `summarize` and `classify`, we offer several optional outputs. I enabled this by using an `--output-base` parameter and internally adding standard extensions, e.g. `.summarized.csv` or `.krona.csv`. I am happy to change this if there's a better design / one more aligned with our other commands. The goal was to enable multiple output formats at once, and allow us to (relatively easily) enable additional output formats for visualization.
- `classify` needs to make some decisions to report the _best_ annotation. I've currently enabled doing this at either a desired rank or given a desired `threshold`. The benefit of the threshold is that we can go up a rank if there's not sufficient support for an annotation at that rank. 
- where do we need to throw errors vs notify and exit?

@ctb @taylorreiter @luizirber thoughts? Anything else needed in this PR?

bluegenes committed:
move threshold arg parsing into cli/utils

bluegenes committed:
Merge branch 'add-taxonomy' of https://github.com/dib-lab/sourmash into add-taxonomy

bluegenes committed:
init changes for multiquery input

bluegenes committed:
use namedtuple for summarized gather results

bluegenes committed:
init update for mult files

bluegenes committed:
adjust for namedtuple output

bluegenes committed:
mods for namedtuple

bluegenes committed:
upd utils

bluegenes committed:
add --from-file to summarize

bluegenes committed:
working multifile summarize

bluegenes committed:
--from-csv to --from-file

ctb commented:
got this error message:
```
% sourmash tax classify

== This is sourmash version 4.1.2. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

Traceback (most recent call last):
  File "/Users/t/miniconda3/envs/py37/bin/sourmash", line 33, in <module>
    sys.exit(load_entry_point('sourmash', 'console_scripts', 'sourmash')())
  File "/Users/t/dev/sourmash/src/sourmash/__main__.py", line 13, in main
    return mainmethod(args)
  File "/Users/t/dev/sourmash/src/sourmash/cli/tax/classify.py", line 67, in main
    return sourmash.tax.__main__.classify(args)
  File "/Users/t/dev/sourmash/src/sourmash/tax/__main__.py", line 109, in classify
    force=args.force)
  File "/Users/t/dev/sourmash/src/sourmash/lca/command_index.py", line 29, in load_taxonomy_assignments
    fp = open(filename, newline='')
TypeError: expected str, bytes or os.PathLike object, not NoneType
```

and similarly
```
% sourmash tax summarize ../genome-grist/outputs.paper/genbank/p8808mo11.x.genbank.gather.csv

== This is sourmash version 4.1.2. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loaded 98 gather results.
Traceback (most recent call last):
  File "/Users/t/miniconda3/envs/py37/bin/sourmash", line 33, in <module>
    sys.exit(load_entry_point('sourmash', 'console_scripts', 'sourmash')())
  File "/Users/t/dev/sourmash/src/sourmash/__main__.py", line 13, in main
    return mainmethod(args)
  File "/Users/t/dev/sourmash/src/sourmash/cli/tax/summarize.py", line 52, in main
    return sourmash.tax.__main__.summarize(args)
  File "/Users/t/dev/sourmash/src/sourmash/tax/__main__.py", line 63, in summarize
    force=args.force)
  File "/Users/t/dev/sourmash/src/sourmash/lca/command_index.py", line 29, in load_taxonomy_assignments
    fp = open(filename, newline='')
TypeError: expected str, bytes or os.PathLike object, not NoneType
```
maybe `--taxonomy-csv` needs a `required=True` in `src/sourmash/cli/tax/summarize.py`?

ctb commented:
I think to try this out I'll need a file that has all of the genbank lineages in it & is properly formatted ;). help?

ctb commented:
> I think to try this out I'll need a file that has all of the genbank lineages in it & is properly formatted ;). help?

never mind! got something working with the latest gtdb `gtdb-rs202.taxonomy.v2.csv`, although it's missing some of the genbank identifiers.

ctb commented:
> This PR adds the top-level framework for a `sourmash tax` subcommand, and focuses on reading in gather results files for summarization or classification.

this description belongs at the top of the PR, I think :).

I got 'classify' and 'summarize' working handily, looks good so far! Will revisit in a bit.

bluegenes committed:
somewhat working classify again

bluegenes committed:
updated classify

bluegenes committed:
finish fixing combine test

bluegenes committed:
make taxonomy_csv required

bluegenes committed:
cleanup

bluegenes committed:
more cleanup

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes committed:
use load_pathlist_from_file

bluegenes committed:
test check_and_load_gather_csvs

bluegenes committed:
properly restrict kwargs with *

bluegenes committed:
allow lineage summary table output from summarize

bluegenes committed:
Merge branch 'latest' into add-taxonomy

ctb committed:
fix merge

bluegenes committed:
require rank for krona, lineage summary output formats

bluegenes committed:
add test for lineage summary output with format_lineage

bluegenes committed:
add docstrings

bluegenes committed:
punt cami to separate PR

bluegenes committed:
raise ValueError on empty gather results

bluegenes committed:
move notify

bluegenes committed:
cleanup

bluegenes committed:
remove tax combine; add tax label

bluegenes committed:
verson of load_taxonomy that strictly uses headers

bluegenes committed:
use new tax fn; enable mult taxonomy inputs

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes committed:
init tax docs

bluegenes committed:
add classification status to classify output

bluegenes committed:
add multi db gather test csv

bluegenes committed:
fix typo

bluegenes committed:
whoops, actually fix

bluegenes committed:
handle accession in lineage csv header

bluegenes committed:
fix line width

bluegenes committed:
return available ranks from load_taxonomy_csv

hehouts committed:
[MRG] add test to confirm failure when summarizing on empty gather (#1560)

* added summarize on empty gather.csv test

* added empty taxonomy csv test

* fix typo

* Update test_tax.py

* trouble shoot tests

* troubleshooting empty gather test

* fixed, maybe? (#1596)

* trying to pull

* cleaned up test_summarize_empty_gather_tax...

* cleaned test_summarize_empty_gather

* fixed test comments

* removed comment

* Update tests/test_tax.py

Co-authored-by: Tessa Pierce Ward <bluegenes@users.noreply.github.com>

* Update tests/test_tax.py

Co-authored-by: Tessa Pierce Ward <bluegenes@users.noreply.github.com>

* updated empty-gather test

Co-authored-by: Tessa Pierce Ward <bluegenes@users.noreply.github.com>
Co-authored-by: C. Titus Brown <titus@idyll.org>

bluegenes committed:
init standardize errs

bluegenes committed:
Merge branch 'add-taxonomy' of https://github.com/sourmash-bio/sourmash into add-taxonomy

bluegenes committed:
add good valueerror for empty lineage csv file

bluegenes committed:
better catch errs in __main__; test all cmds: empty gather, lineage files

bluegenes committed:
check available ranks, bad gather headers, empty gather, etc

bluegenes committed:
emit one (and only one) warning per 100% match

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes committed:
add all functions to __all__ ...is this desired?

bluegenes commented:
ready for review! @ctb @luizirber @taylorreiter

bluegenes committed:
Merge branch 'latest' into add-taxonomy

luizirber commented on doc/command-line.md:
```diff
@@ -411,7 +421,94 @@ This combination of commands ensures that the more time- and
+- `summary` is the default output format. This outputs a `csv` with lineage
+summarization for each taxonomic rank. This output currently consists of four
+columns, `query_name,rank,fraction,lineage`, where `fraction` is the  fraction
+of the query matched to the reported rank and lineage.  - `krona` format is a
```
```suggestion
of the query matched to the reported rank and lineage.
- `krona` format is a
```

luizirber commented on doc/command-line.md:
```diff
@@ -411,7 +421,94 @@ This combination of commands ensures that the more time- and
+`fraction` is the fraction of the query matched to the reported rank and
+lineage. The remaining columns are `superkingdom`, `phylum`, .. etc down to the
+rank used for summarization. This output can be used directly for summary
+visualization.  - `lineage_summary` - the lineage summary format is most useful
```
```suggestion
visualization.
- `lineage_summary` - the lineage summary format is most useful
```

luizirber commented on doc/command-line.md:
```diff
@@ -411,7 +421,94 @@ This combination of commands ensures that the more time- and
+- `summary` is the default output format. This outputs a `csv` with lineage
+summarization for each taxonomic rank. This output currently consists of four
+columns, `query_name,rank,fraction,lineage`, where `fraction` is the  fraction
+of the query matched to the reported rank and lineage.  - `krona` format is a
```
```suggestion
of the query matched to the reported rank and lineage.
- `krona` format is a
```

bluegenes committed:
change cli for better arg parsing; add examples to summarize docs

bluegenes committed:
Merge branch 'add-taxonomy' of https://github.com/sourmash-bio/sourmash into add-taxonomy

bluegenes committed:
add mixed strain classify example and assoc data

bluegenes committed:
doc formatting

bluegenes committed:
more doc formatting

bluegenes committed:
more tiny reformatting

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+ information, see [databases](databases.md).
+
+These commands rely upon the fact that `gather` results are non-overlapping: the
+ fraction match for gather on each query will be between 0 (no database matches)
```
it would be good to define fraction match here. Do you mean `f_unique_to_query`? You should consider referring to classifying-signatures.html.

ctb replied:
added issue to update `classifying-signatures` at some point :) https://github.com/sourmash-bio/sourmash/issues/1609

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+ reference databases.
+
+
+### `sourmash tax summarize` (for summarizing metagenomes)
```
```suggestion
### `sourmash tax summarize` - summarize metagenome content from `gather` results
```
(this keeps it in the same format as other content in this doc)

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+### `sourmash tax summarize` (for summarizing metagenomes)
+
+`sourmash tax summarize` - for each gather query, summarize gather results by
```
```suggestion
`sourmash tax summarize` summarizes gather results for each query by
```

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+#### `summary` output format
+
+`summary` is the default output format. This outputs a `csv` with lineage
```
maybe make this `csv_summary`? (soft suggestion)

Thoughts:
* It kind of doesn't matter how long the string is, if it's default
* we've got a lot of `summary' strings floating around already :)
* it's nice to indicate that it's parseable

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+To generate `lineage_summary`, we add `--output-format lineage_summary` to the summarize
+ command, and need to specify a rank to summarize. Here's the command for reporting
+ `lineage_summary` for two queries (HSMA33MX, NTP_ADD_ME) summary at `species` level.
```
note: `NTP_ADD_ME`

ctb commented on doc/command-line.md:
````diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+example `lineage_summary`:
+
+```
+lineage	HSMA33MX ADD_ME
````
note: `ADD_ME`

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+ `--output-format` argument, e.g. `--output-format summary krona lineage_summary`
+
+
+### `sourmash tax classify` (for classifying genomes)
```
```suggestion
### `sourmash tax classify` - classify a genome using `gather` results
```
again, in keeping with style of document so far

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+### `sourmash tax classify` (for classifying genomes)
+
+`sourmash tax classify` - for each gather query, report likely classification
```
```suggestion
`sourmash tax classify` reports likely classification for each query,
```

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+> Here, `f_match` shows that independently, both strains match ~65% percent of
+ this mixed query. The `f_unique_weighted` column has the results of gather-style
+ decomposition. So the OS223 strain had a slightly higher `f_match` (66%), so it
```
```suggestion
 decomposition. As the OS223 strain had a slightly higher `f_match` (66%), it
```

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+> Here, `f_match` shows that independently, both strains match ~65% percent of
+ this mixed query. The `f_unique_weighted` column has the results of gather-style
+ decomposition. So the OS223 strain had a slightly higher `f_match` (66%), so it
+ was the first match. Then, the matching sequence was removed from the query. The
```
```suggestion
 was the first match. The
```
I think the final sentence is sufficient ("the remaining 33%") to indicate that the matching sequence was removed from the query.

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+There are two possible output formats, `summary` and `krona`.
+
+#### `summary` output format
```
maybe make this `csv_summary`? (in line with same suggestion made above for `summarize`)

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+To produce multiple output types from the same command, add the types into the
+ `--output-format` argument, e.g. `--output-format summary krona`.
+ **Note that for classify, specifying `--rank`, as needed for `krona`, forces
```
Maybe rewrite as:
"Note here that specifying `--rank` forces classification by rank rather than by containment threshold, i.e. the option XYZ is implied."
See above for request to add whatever option XYZ is into the docs :)

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+ matched to one species, and an additional 5% matched to a different species in
+ the same genus, the genus-level classification will be reported.
+
+Optionally, `classify` can instead report classifications at a desired `rank`,
```
Is this a command-line option? If so, what is it / how do I turn it on? (See comment about 'XYZ' below, also.)

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+ (default=0.1), the `status` column will contain `below_threshold`.
+
+
+### `sourmash tax label` (for labeling gather results)
```
```suggestion
### `sourmash tax label` - annotates gather output with taxonomy
```

also, could be `annotate` rather than `label` (soft suggestion). `label` doesn't immediately jump out to me as meaningful here.

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+
+### `sourmash tax label` (for labeling gather results)
+
+`sourmash tax label` - for any gather results, add a column with taxonomic
```
```suggestion
`sourmash tax label` adds a column with taxonomic 
```

ctb commented on doc/command-line.md:
```diff
@@ -411,7 +421,252 @@ This combination of commands ensures that the more time- and
+### `sourmash tax label` (for labeling gather results)
+
+`sourmash tax label` - for any gather results, add a column with taxonomic
+ lineage information for each database match. Do not summarize or classify. Note
```
```suggestion
 lineage information for each database match to gather output. Do not summarize or classify. Note
```

ctb commented on src/sourmash/cli/tax/__init__.py:
```diff
@@ -0,0 +1,32 @@
+        helpstring = 'sourmash tax {op:s} --help'.format(op=subcmd)
+        desc += '        {hs:33s} {ds:s}\n'.format(hs=helpstring, ds=docstring)
+    s = subparser.add_subparsers(
+        title='Integrate taxonomy information', dest='subcmd', metavar='subcmd', help=SUPPRESS,
```
does this now get confusing in contrast to `lca`? Do we want to put `gather` in the description somewhere?

(soft suggestion, just asking)

I think this results in the following output when running `sourmash`:

```

    Integrate taxonomy information
        sourmash tax --help

```
in which case adding `based on 'gather' results` could be good. I'll put it in as a suggestion.

ctb replied:
```suggestion
        title="Integrate taxonomy information based on 'gather' results", dest='subcmd', metavar='subcmd', help=SUPPRESS,
```

ctb commented on src/sourmash/cli/tax/classify.py:
```diff
@@ -0,0 +1,68 @@
+from sourmash.cli.utils import add_threshold_arg
+
+def subparser(subparsers):
+    subparser = subparsers.add_parser('classify')
```
please add `help=...` information here, so that `sourmash tax classify` output is more informative -

currently, it's
```
% sourmash tax classify
usage:  classify [-h] [-g [GATHER_CSV [GATHER_CSV ...]]] [--from-file FILE]
                 [-q] -t FILE [FILE ...] [-o OUTPUT_BASE]
                 [-r {strain,species,genus,family,order,class,phylum,superkingdom}]
                 [--keep-full-identifiers] [--keep-identifier-versions]
                 [--fail-on-missing-taxonomy]
                 [--output-format {summary,krona} [{summary,krona} ...]] [-f]
                 [--containment-threshold CONTAINMENT_THRESHOLD]
 classify: error: the following arguments are required: -t/--taxonomy-csv

ctb commented on src/sourmash/cli/tax/classify.py:
```diff
@@ -0,0 +1,68 @@
+    subparser = subparsers.add_parser('classify')
+    subparser.add_argument(
+        '-g', '--gather-csv', nargs='*', default = [],
+        help='csvs from sourmash gather'
```
```suggestion
        help='CSVs output by sourmash gather for this sample'
```
(or something similar)

ctb commented on src/sourmash/cli/tax/classify.py:
```diff
@@ -0,0 +1,68 @@
+    )
+    subparser.add_argument(
+        '--from-file',  metavar='FILE', default=None,
+        help='input many gather results as a text file, with one gather csv per line'
```
```suggestion
        help='input many gather results as a text file, with one gather CSV per line'
```

ctb commented on src/sourmash/cli/tax/classify.py:
```diff
@@ -0,0 +1,68 @@
+    subparser.add_argument(
+        '-t', '--taxonomy-csv',  metavar='FILE',
+        nargs='+', required=True,
+        help='database lineages csv'
```
```suggestion
        help='database lineages CSV'
```

ctb commented on src/sourmash/cli/tax/classify.py:
```diff
@@ -0,0 +1,68 @@
+    )
+    subparser.add_argument(
+        '-r', '--rank', choices=['strain','species', 'genus', 'family', 'order', 'class', 'phylum', 'superkingdom'],
+        help='Summarize genome taxonomy at this rank and above. Note that the taxonomy csv must contain lineage information at this rank.'
```
```suggestion
        help='Summarize genome taxonomy at this rank and above. Note that the taxonomy CSV must contain lineage information at this rank.'
```

ctb commented on src/sourmash/cli/tax/classify.py:
```diff
@@ -0,0 +1,68 @@
+def main(args):
+    import sourmash
+    if not args.gather_csv and not args.from_file:
+        raise ValueError(f"No gather csvs found! Please input via `-g` or `--from-file`.")
```
```suggestion
        raise ValueError(f"No gather CSVs found! Please input via `-g` or `--from-file`.")
```

ctb commented on src/sourmash/cli/tax/label.py:
```diff
@@ -0,0 +1,51 @@
+    subparser = subparsers.add_parser('label')
+    subparser.add_argument(
+        '-g', '--gather-csv', nargs='*', default = [],
+        help='csvs from sourmash gather'
```
```suggestion
        help='CSV output files from sourmash gather'
```
or something similar

ctb commented on src/sourmash/cli/tax/label.py:
```diff
@@ -0,0 +1,51 @@
+
+
+def subparser(subparsers):
+    subparser = subparsers.add_parser('label')
```
please add `help=...` here, too.

ctb commented on src/sourmash/cli/tax/label.py:
```diff
@@ -0,0 +1,51 @@
+    )
+    subparser.add_argument(
+        '--from-file',  metavar='FILE', default=None,
+        help='input many gather results as a text file, with one gather csv per line'
```
```suggestion
        help='input many gather results as a text file, with one gather CSV per line'
```

ctb commented on src/sourmash/cli/tax/label.py:
```diff
@@ -0,0 +1,51 @@
+    subparser.add_argument(
+        '-t', '--taxonomy-csv', metavar='FILE',
+        nargs="+", required=True,
+        help='database lineages csv'
```
```suggestion
        help='database lineages CSV'
```

ctb commented on src/sourmash/cli/tax/label.py:
```diff
@@ -0,0 +1,51 @@
+    )
+    subparser.add_argument(
+        '-o', '--output-dir', default= "",
+        help='directory for output csv(s)'
```
```suggestion
        help='directory for output files'
```

ctb replied:
(or, 'CSV files')

ctb commented on src/sourmash/cli/tax/label.py:
```diff
@@ -0,0 +1,51 @@
+def main(args):
+    import sourmash
+    if not args.gather_csv and not args.from_file:
+        raise ValueError(f"No gather csvs found! Please input via `-g` or `--from-file`.")
```
```suggestion
        raise ValueError(f"No gather CSVs found! Please input via `-g` or `--from-file`.")
```

ctb commented on src/sourmash/cli/tax/summarize.py:
```diff
@@ -0,0 +1,65 @@
+
+
+def subparser(subparsers):
+    subparser = subparsers.add_parser('summarize')
```
please add `help=...`

ctb commented on src/sourmash/cli/tax/summarize.py:
```diff
@@ -0,0 +1,65 @@
+    subparser = subparsers.add_parser('summarize')
+    subparser.add_argument(
+        '-g', '--gather-csv', nargs='*', default = [],
+        help='csvs from sourmash gather'
```
```suggestion
        help='CSVs from sourmash gather'
```

ctb commented on src/sourmash/cli/tax/summarize.py:
```diff
@@ -0,0 +1,65 @@
+    )
+    subparser.add_argument(
+        '--from-file',  metavar='FILE', default = None,
+        help='input many gather results as a text file, with one gather csv per line'
```
```suggestion
        help='input many gather results as a text file, with one gather CSV per line'
```

ctb commented on src/sourmash/cli/tax/summarize.py:
```diff
@@ -0,0 +1,65 @@
+    subparser.add_argument(
+        '-t', '--taxonomy-csv', metavar='FILE',
+        nargs='+', required=True,
+        help='database lineages csv'
```
```suggestion
        help='database lineages CSV'
```

ctb commented on src/sourmash/cli/tax/summarize.py:
```diff
@@ -0,0 +1,65 @@
+    )
+    subparser.add_argument(
+        '-r', '--rank', choices=['strain','species', 'genus', 'family', 'order', 'class', 'phylum', 'superkingdom'],
+        help='For non-default output formats: Summarize genome taxonomy at this rank and above. Note that the taxonomy csv must contain lineage information at this rank.'
```
```suggestion
        help='For non-default output formats: Summarize genome taxonomy at this rank and above. Note that the taxonomy CSV must contain lineage information at this rank.'
```

ctb commented on src/sourmash/cli/tax/summarize.py:
```diff
@@ -0,0 +1,65 @@
+def main(args):
+    import sourmash
+    if not args.gather_csv and not args.from_file:
+        raise ValueError(f"No gather csvs found! Please input via `-g` or `--from-file`.")
```
```suggestion
        raise ValueError(f"No gather CSVs found! Please input via `-g` or `--from-file`.")
```

ctb commented on src/sourmash/cli/utils.py:
```diff
@@ -50,11 +50,34 @@ def add_ksize_arg(parser, default=31):
         '--picklist', default=None,
         help="select signatures based on a picklist, i.e. 'file.csv:colname:coltype'"
     )
+    
```
```suggestion
```
that's a lot of extra whitespace there, pardner!

ctb commented on src/sourmash/cli/utils.py:
```diff
@@ -50,11 +50,34 @@ def add_ksize_arg(parser, default=31):
         help="select signatures based on a picklist, i.e. 'file.csv:colname:coltype'"
     )
+    
+    
```
```suggestion
```
remove whitespace round 2

ctb commented on src/sourmash/cli/utils.py:
```diff
@@ -50,11 +50,34 @@ def add_ksize_arg(parser, default=31):
+    return f
+
+
+def add_threshold_arg(parser, default=0.1):
```
please make this `add_tax_threshold_arg` or something more specific to this PR, thanks!

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+"""
+import sys
+import csv
+import json
```
this is imported but unused.

(please run `pyflakes src/sourmash/tax` and fix all the relevant warnings; there's a lot of unused stuff in here)

bluegenes replied:
done! sidenote, hadn't done this before and it was super neat and helpful

ctb replied:
right? I have started to use it more habitually ;).

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+            this_tax_assign, _, avail_ranks = tax_utils.load_taxonomy_csv(tax_csv, split_identifiers=not args.keep_full_identifiers,
+                                              keep_identifier_versions = args.keep_identifier_versions,
+                                              force=args.force)
+            # maybe check for overlapping tax assignments? rn later ones will override earlier ones
```
```suggestion
            # maybe check for overlapping tax assignments? currently, later ones will override earlier ones
```

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+            error(str(exc))
+
+    if not tax_assign:
+        error(f'No taxonomic assignments loaded from {args.taxonomy_csv}. Exiting.')
```
```suggestion
        error(f'ERROR: No taxonomic assignments loaded from {args.taxonomy_csv}. Exiting.')
```

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+        sys.exit(-1)
+
+    if args.rank and args.rank not in available_ranks:
+        error(f"No taxonomic information provided for rank {args.rank}: cannot summarize at this rank")
```
```suggestion
        error(f"ERROR: No taxonomic information provided for rank {args.rank}: cannot summarize at this rank")
```

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+                                                                seen_perfect = seen_perfect)
+
+    # write summarized output csv
+    if "summary" in args.output_format:
```
idle curiousity: what happens if there are no output formats specified? warning? error?

bluegenes replied:
the default is this `summary` (or tbd, `csv_summary`) format, so `args.output_format` should never be empty

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+            tax_assign.update(this_tax_assign)
+            available_ranks.update(set(avail_ranks))
+        except ValueError as exc:
+            error(str(exc))
```
should this exit afterwards? usually when I use `error(...)` it means that (barring a `--force`) we should exit.

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+        sys.exit(-1)
+
+    if args.rank and args.rank not in available_ranks:
+        error(f"No taxonomic information provided for rank {args.rank}: cannot classify at this rank")
```
```suggestion
        error(f"ERROR: No taxonomic information provided for rank {args.rank}: cannot classify at this rank")
```

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+            error(str(exc))
+
+    if not tax_assign:
+        error(f'No taxonomic assignments loaded from {args.taxonomy_csv}. Exiting.')
```
```suggestion
        error(f'ERROR: No taxonomic assignments loaded from {args.taxonomy_csv}. Exiting.')
```

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+                    continue
+                if fraction <= args.containment_threshold:
+                    status="below_threshold"
+                    notify(f"WARNING: classifying at desired rank {args.rank} does not meet containment threshold {args.containment_threshold}")
```
does this give enough information to figure out what entry is problematic? maybe include `query_name` in the error output?

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+                    notify(f"WARNING: classifying at desired rank {args.rank} does not meet containment threshold {args.containment_threshold}")
+                else:
+                    status="match"
+                classifications[args.rank].append((query_name, status, rank, fraction, lineage))
```
maybe make this a namedtuple? `(query_name, status, rank, fraction, lineage)`?

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+    if "summary" in args.output_format:
+        summary_outfile = make_outfile(args.output_base, ".classifications.csv")
+        with FileOutputCSV(summary_outfile) as out_fp:
+            #tax_utils.write_summary(classifications, out_fp)
```
?

ctb commented on src/sourmash/tax/__main__.py:
```diff
@@ -0,0 +1,300 @@
+        tax_assign.update(this_tax_assign)
+
+    if not tax_assign:
+        error(f'No taxonomic assignments loaded from {args.taxonomy_csv}. Exiting.')
```
```suggestion
        error(f'ERROR: No taxonomic assignments loaded from {args.taxonomy_csv}. Exiting.')
```

ctb reviewed (changes requested):
Looking good!

ctb commented:
Some larger notes from my review --

I think `summarize` and `classify` are nice names but just to be clear I invented them out of whole cloth (as far as I recall) when I built the `lca` module. So we could go with something else if we wanted (and alias `lca` to match! or alias in both directions!).

I was thinking maybe:
* `sourmash tax metagenome` 
* `sourmash tax genome`

which seem very clear to me.

Separately, I wonder if we could or should have these take `search`/`gather`/`prefetch` style command lines?

I addressed this over in #1515 with this comment:
>the --query and --db style to take multiple arguments is clumsy when you have one query vs 1 db, but is really nice when you have multiple. dunno what to do here, but suggest being consistent within the subcommand.

I'm thinking that rather than 
```
sourmash tax classify --gather-csv XYZ.csv --taxonomy-csv GTDB.csv
```
we could go with
```
sourmash tax classify XYZ.csv --taxonomy-csv GTDB.csv
```
which would then nicely permit 
```
sourmash tax classify gather/*.csv --taxonomy-csv GTDB.csv
```
Not a big deal either way, but wanted to raise the question!

bluegenes committed:
Apply suggestions from code review

Co-authored-by: C. Titus Brown <titus@idyll.org>

bluegenes committed:
add usage info for each subcommand; upd docs

bluegenes committed:
minor cleanup and more informative warning

bluegenes commented:
> Some larger notes from my review --
> 
> I think `summarize` and `classify` are nice names but just to be clear I invented them out of whole cloth (as far as I recall) when I built the `lca` module. So we could go with something else if we wanted (and alias `lca` to match! or alias in both directions!).
> 
> I was thinking maybe:
> 
> * `sourmash tax metagenome`
> * `sourmash tax genome`
> 
> which seem very clear to me.
> 

I would love @taylorreiter's thoughts on these ^! They seem clear to me as well, but so do `summarize` and `classify`. I suppose `metagenome` and `genome` do make the functional distinction much clearer... 


> Separately, I wonder if we could or should have these take `search`/`gather`/`prefetch` style command lines?
> 
> I addressed this over in #1515 with this comment:
> 
> > the --query and --db style to take multiple arguments is clumsy when you have one query vs 1 db, but is really nice when you have multiple. dunno what to do here, but suggest being consistent within the subcommand.
> 
> I'm thinking that rather than
> 
> ```
> sourmash tax classify --gather-csv XYZ.csv --taxonomy-csv GTDB.csv
> ```
> 
> we could go with
> ```
> sourmash tax classify XYZ.csv --taxonomy-csv GTDB.csv
> 
> ```
> which would then nicely permit 
> ```
> sourmash tax classify gather/*.csv --taxonomy-csv GTDB.csv
> ```
> Not a big deal either way, but wanted to raise the question!

This was how I had it written! But enabling multiple gather csvs AND multiple taxonomy csvs complicated things, essentially requiring the positional gather csv's to come **before** the taxonomy argument, otherwise the taxonomy arg gobbled up the gather csv's (thanks to taylor for finding/breaking this immediately on friday :). So in https://github.com/sourmash-bio/sourmash/pull/1543/commits/e8f0b68958a04f9755c2cf683df3693e2060190e, I changed to this arg style. If you have a different workaround, I'm all ears!

ctb commented:
> > Some larger notes from my review --
> > I think `summarize` and `classify` are nice names but just to be clear I invented them out of whole cloth (as far as I recall) when I built the `lca` module. So we could go with something else if we wanted (and alias `lca` to match! or alias in both directions!).
> > I was thinking maybe:
> > 
> > * `sourmash tax metagenome`
> > * `sourmash tax genome`
> > 
> > which seem very clear to me.
> 
> I would love @taylorreiter's thoughts on these ^! They seem clear to me as well, but so do `summarize` and `classify`. I suppose `metagenome` and `genome` do make the functional distinction much clearer...

`mixture` and `isolate` serve as good reminders that clarity is not always better (as in: I hate them, even though they are very clear :).

ctb commented:
> This was how I had it written! But enabling multiple gather csvs AND multiple taxonomy csvs complicated things, essentially requiring the positional gather csv's to come **before** the taxonomy argument, otherwise the taxonomy arg gobbled up the gather csv's (thanks to taylor for finding/breaking this immediately on friday :). So in [e8f0b68](https://github.com/sourmash-bio/sourmash/commit/e8f0b68958a04f9755c2cf683df3693e2060190e), I changed to this arg style. If you have a different workaround, I'm all ears!

Hah, good to know. No objections.

So... thinking more on this... a different question arises: should we allow multiple gather CSVs for the same sample? I think the answer should be "no", because in most situations that will break the assumptions of no overlap.

(And, with prefetch and picklists, it's easy enough to run prefetch against multiple databases, save the prefetch CSVs, and then use them as picklists to do a gather against multiple databases.)

One clear use case here is when you want to use GTDB, NCBI all, and all phages from a private database, and don't want to run a single gather...

bluegenes commented:
> So... thinking more on this... a different question arises: should we allow multiple gather CSVs for the same sample? I think the answer should be "no", because in most situations that will break the assumptions of no overlap.
> 
> (And, with prefetch and picklists, it's easy enough to run prefetch against multiple databases, save the prefetch CSVs, and then use them as picklists to do a gather against multiple databases.)

Ah, I'm not doing any summarization _across_ those multiple gather CSV's, only within each gather CSV. The reason to enable multiple gather CSVs is that it enables the `lineage_summary` format (table of the fraction summarized for each lineage in each sample/gather query) to come directly out of this command rather than, say, a `combine` command.

> One clear use case here is when you want to use GTDB, NCBI all, and all phages from a private database, and don't want to run a single gather...

Don't we need to run just a single gather against all databases at once in order to properly allow summarization?

Thinking more about changing command names, I do think that the presence of `lca summarize`, `lca classify`,`tax summarize` and `tax classify` will be a bit confusing, and might be another reason to switch to `genome`/`metagenome`. Is the plan to keep supporting those `lca` commands? I might take a stab at updating the `classifying-signatures` docs -- if we're planning on supporting both, that might be the place to discuss the different strategies.

As for `label` --> `annotate`, I'd be ok with this. I think `annotate` still gives me gene function connotations, but `tax annotate` does make sense!

ctb commented:
> > So... thinking more on this... a different question arises: should we allow multiple gather CSVs for the same sample? I think the answer should be "no", because in most situations that will break the assumptions of no overlap.
> > (And, with prefetch and picklists, it's easy enough to run prefetch against multiple databases, save the prefetch CSVs, and then use them as picklists to do a gather against multiple databases.)
> 
> Ah, I'm not doing any summarization _across_ those multiple gather CSV's, only within each gather CSV. The reason to enable multiple gather CSVs is that it enables the `lineage_summary` format (table of the fraction summarized for each lineage in each sample/gather query) to come directly out of this command rather than, say, a `combine` command.

OK - makes sense. Do you raise an alert if the same sample name shows up in multiple gather CSVs?
 
> > One clear use case here is when you want to use GTDB, NCBI all, and all phages from a private database, and don't want to run a single gather...
> 
> Don't we need to run just a single gather against all databases at once in order to properly allow summarization?

right - independent prefetch -> combined gather using prefetch output+picklists, is my suggested solution here.

> Thinking more about changing command names, I do think that the presence of `lca summarize`, `lca classify`,`tax summarize` and `tax classify` will be a bit confusing, and might be another reason to switch to `genome`/`metagenome`. Is the plan to keep supporting those `lca` commands? I might take a stab at updating the `classifying-signatures` docs -- if we're planning on supporting both, that might be the place to discuss the different strategies.

I feel like I discussed this somewhere, but can't find it at the moment! But also I don't remember coming to a conclusion? Maybe it was all just in my head :).

Hot takes:

I think the lca functionality is still useful, for sure until we've really proven out the gather taxonomy stuff, but perhaps even after that?

We should absolutely update classifying-signatures, but it can wait until after a release. Let's just make sure to make an issue and get to it sooner rather than later ;).

I'm good with your idea of changing to use `tax metagenome` and `tax genome` so that it doesn't overlap with `lca` terminology.
I'm also good with aliasing the `lca summarize` and `lca classify`to match `tax` (e.g. `metagenome`, `genome`), and/or changing the documents to reflect new names, OR just leaving it as an easter egg. I guess I'd lean slightly towards (a) using `tax metagenome` and `tax genome`, and (b) doing all the aliases in both directions and writing tests, and (c) not documenting the aliases 😁 

> As for `label` --> `annotate`, I'd be ok with this. I think `annotate` still gives me gene function connotations, but `tax annotate` does make sense!

👍

bluegenes commented:
> OK - makes sense. Do you raise an alert if the same sample name shows up in multiple gather CSVs?

Whoops, used to, but I guess I lost it when switching to multiple gather CSVs. Adding now -- do we want to ignore duplicates by default (emitting a warning) or only ignore with `--force`? I'm leaning the latter...

ctb commented:
> > OK - makes sense. Do you raise an alert if the same sample name shows up in multiple gather CSVs?
> 
> Whoops, used to, but I guess I lost it when switching to multiple gather CSVs. Adding now -- do we want to ignore duplicates by default (emitting a warning) or only ignore with `--force`? I'm leaning the latter...

I look forward to informing you why whatever it is you chose is problematic!

taylorreiter commented:
Been pondering words...

> >    I think summarize and classify are nice names but just to be clear I invented them out of whole cloth (as far as I recall) when I built the lca module. So we could go with something else if we wanted (and alias lca to match! or alias in both directions!).

>>    I was thinking maybe:
```
        sourmash tax metagenome
        sourmash tax genome
```
>>    which seem very clear to me.

> I would love @taylorreiter's thoughts on these ^! They seem clear to me as well, but so do summarize and classify. I suppose metagenome and genome do make the functional distinction much clearer...

I think `metagenome` and `genome` is ~good. I don't think I like `mixture` and `isolate` that much...Isolate has a particular use-case, and a genome-bin isn't an isolate (it isn't really a genome either, but since databases refer to is as such, shrug). 

So this all sounds great:

>I'm good with your idea of changing to use tax metagenome and tax genome so that it doesn't overlap with lca terminology.
> I'm also good with aliasing the lca summarize and lca classifyto match tax (e.g. metagenome, genome), and/or changing the documents to reflect new names, OR just leaving it as an easter egg. I guess I'd lean slightly towards (a) using tax metagenome and tax genome, and (b) doing all the aliases in both directions and writing tests, and (c) not documenting the aliases 😁

As for adding lineages to a gather spreadsheet, 

>    As for label --> annotate, I'd be ok with this. I think `annotate` still gives me gene function connotations, but `tax annotate` does make sense!

I'm with @bluegenes that `annotate` has other connotations tide to it. `add_lineage` or `label_lineage` aren't very snappy, but they are very descriptive. That's the direction i'm currently leaning.

bluegenes committed:
add checks for duplicated queries, better output for duplicated filenames; label --> annotate

bluegenes committed:
better catch/print for valueerrors; pyflakes fixes

bluegenes committed:
summary --> csv_summary

bluegenes committed:
upd docs

bluegenes committed:
two-sample lineage summary example

bluegenes committed:
rename commands

bluegenes committed:
upd function names too

bluegenes committed:
Merge branch 'latest' into add-taxonomy

bluegenes commented on doc/command-line.md:
```diff
@@ -411,7 +421,268 @@ This combination of commands ensures that the more time- and
+
+These commands rely upon the fact that `gather` provides both the total
+ fraction of the query matched to each database matched, as well as a
+ non-overlapping `f_unique_weighted` which is the fraction of the query
```
@ctb should I use `f_unique_to_query` here instead?

ctb commented:
yes, I think so

bluegenes committed:
minor doc upds

bluegenes commented:
> yes, I think so

ok, will do, but I think we need to clarify / extend docs on how abundances are used!

bluegenes committed:
use f_unique_to_query

ctb commented:
Ran through some real-ish data - here are the results! I'll make a specific checklist in the PR review.

(content from https://hackmd.io/KctUsXsLTWGdCuqXS9x4nw)

# taxonomy PR tryout!

## SRR606249 metagenome ('podar')

### run a prefetch of SRR606249 against GTDB genomic reps:
```
% sourmash prefetch SRR606249-k31.sig  gtdb-r202.genomic-reps.k31.zip -o SRR606249-k31.x.gtdb.prefetch.csv --save-matches SRR606249-k31.x.gtdb.prefetch.zip
```
This finds 780 matches.

### run a gather of SRR606249 against GTDB genomic reps:
```
% sourmash gather SRR606249-k31.sig SRR606249-k31.x.gtdb.prefetch.zip -o SRR606249-k31.x.gtdb.gather.csv
```
This finds 84 matches.

### run `tax annotate`

```
% sourmash tax annotate -g SRR606249-k31.x.gtdb.gather.csv  -t gtdb-rs202.taxonomy.v2.csv 
```

which gives:

```
== This is sourmash version 4.1.3.dev7+g0814bcc5. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loaded 84 gather results.
of 84, missed 0 lineage assignments.
loaded results from 1 gather CSVs
```

CTB: I think some output message here would be nice - what file(s) is it outputting? Looks like it's `SRR606249-k31.x.gtdb.gather.with-lineages.csv`.

### run `tax metagenome`

```
% sourmash tax metagenome -g SRR606249-k31.x.gtdb.gather.with-lineages.csv -t gtdb-rs202.taxonomy.v2.csv -o BASE
```

works great :)

CTB: I think some output message here would be nice - what file(s) is it outputting? (In this case, `BASE.summarize.csv`)

## 'Aciduliprofundum' query

### run a prefetch of podar-ref/1.fa.sig against GTDB

```
% sourmash prefetch podar-ref/1.fa.sig gtdb-r202.genomic-reps.k31.zip -o 1.x.gtdb.prefetch.csv --save-matches 1.x.gtdb.prefetch.zip -k 31
```

1 match.

### run a gather

```
% sourmash gather podar-ref/1.fa.sig 1.x.gtdb.prefetch.zip  -o 1.x.gtdb.gather.csv
```

1 match.

### run `tax annotate` incorrectly

By mistake, I ran
```
% sourmash tax annotate -t 1.x.gtdb.gather.csv -g gtdb-rs202.taxonomy.v2.csv 
```
and got the error message `ERROR: No taxonomic identifiers found.`

CTB: let's make an issue to detect if -g and -t were switched and provide helpful error output :). it's a "good next issue" I think.

### run `tax annotate` correctly

```
% sourmash tax annotate -g 1.x.gtdb.gather.csv -t gtdb-rs202.taxonomy.v2.csv 
```

works!

### run `tax genome`

```
% sourmash tax genome -g 1.x.gtdb.gather.csv -t gtdb-rs202.taxonomy.v2.csv 
```

gives:

```
== This is sourmash version 4.1.3.dev7+g0814bcc5. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loaded 1 gather results.
of 1, missed 0 lineage assignments.
loaded results from 1 gather CSVs
WARNING: 100% match! Is query CP001941.1 Aciduliprofundum boonei T469, complete genome identical to its database match, GCF_000025665?
query_name,status,rank,fraction,lineage
"CP001941.1 Aciduliprofundum boonei T469, complete genome",match,species,1.000,d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Aciduliprofundales;f__Aciduliprofundaceae;g__Aciduliprofundum;s__Aciduliprofundum boonei
```

CTB: please add quotes around the signature name after "Is query ... identical?"

Looking at the output, I see:

```
query_name,status,rank,fraction,lineage
"CP001941.1 Aciduliprofundum boonei T469, complete genome",match,species,1.000,d__Archaea;p__Thermoplasmatota;c__Thermoplasmata;o__Aciduliprofundales;f__Aciduliprofundaceae;g__Aciduliprofundum;s__Aciduliprofundum boonei
```

CTB: I think for both `metagenome` and `genome` we should add two more columns, `query_md5` and `query_filename`, to the output.

ctb reviewed (changes requested):
Looking good! A few small things to fix --

- [x] add output messages indicating which file(s) are being produced, for `tax annotate`
- [x] add output messages indicating which file(s) are being produced, for `tax genome`
- [x] add output messages indicating which file(s) are being produced, for `tax metagenome`
- [x] create a "good next issue" to detect if -g and -t were switched and provide helpful error output
- [x] in `tax genome`, please add quotes around the signature name in "Is query {name} identical?"
- [x] add two more columns, query_md5 and query_filename, to the `csv_summary` output from `tax metagenome`
- [x] add two more columns, query_md5 and query_filename, to the `csv_summary` output from `tax genome`

ctb commented:
> > yes, I think so
> 
> ok, will do, but I think we need to clarify / extend docs on how abundances are used!

Hot take: the docs [here](https://sourmash.readthedocs.io/en/latest/classifying-signatures.html#abundance-weighting) (`classifying-signatures.md`) are good enough, and what we really need is someone to do some real-world validation and then feed that experience back into the docs.

bluegenes committed:
add output dir; add file output notifications

bluegenes committed:
add cols to summary outputs; adjust accordingly

bluegenes committed:
Merge branch 'latest' into add-taxonomy

ctb reviewed (approved):
🎉

ctb commented:
huh, tests aren't passing tho?

bluegenes commented:
> huh, tests aren't passing tho?
> 
> 

They keep getting canceled by server. I've hit rerun - hopefully they'll run this time.

ctb committed:
trigger GitHub actions

ctb committed:
Merge branch 'add-taxonomy' of github.com:dib-lab/sourmash into add-taxonomy
